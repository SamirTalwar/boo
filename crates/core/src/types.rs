//! A representation of a value's type.
//!
//! Used for type-checking and valid program synthesis.

use std::borrow::Borrow;
use std::fmt::Display;
use std::sync::Arc;

/// An opaque wrapper around a type.
pub trait TypeRef: From<Type<Self>> + Display + Sized {}

/// A simple type wrapper that allows for cycles.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Monotype(pub Arc<Type<Self>>);

impl AsRef<Type<Self>> for Monotype {
    fn as_ref(&self) -> &Type<Self> {
        self.0.as_ref()
    }
}

impl From<Type<Self>> for Monotype {
    fn from(value: Type<Self>) -> Self {
        Self(Arc::new(value))
    }
}

impl Display for Monotype {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.0.fmt(f)
    }
}

impl TypeRef for Monotype {}

/// A type bound by forall quantifiers.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Polytype {
    pub quantifiers: Vec<TypeVariable>,
    pub mono: Monotype,
}

impl Polytype {
    pub fn unquantified(mono: Monotype) -> Self {
        Polytype {
            quantifiers: vec![],
            mono,
        }
    }
}

impl Display for Polytype {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if self.quantifiers.is_empty() {
            write!(f, "{}", self.mono)
        } else {
            write!(f, "âˆ€")?;
            for quantifier in self.quantifiers.iter() {
                write!(f, " {quantifier}")?;
            }
            write!(f, ". ")?;
            write!(f, "{}", self.mono)
        }
    }
}

/// The set of types.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type<Outer: TypeRef> {
    Integer,
    Function { parameter: Outer, body: Outer },
    Variable(TypeVariable),
}

impl<Outer: TypeRef> Type<Outer> {
    pub fn transform<NewOuter: TypeRef>(self, f: impl Fn(Outer) -> NewOuter) -> Type<NewOuter> {
        match self {
            Type::Integer => Type::Integer,
            Type::Function { parameter, body } => Type::Function {
                parameter: f(parameter),
                body: f(body),
            },
            Type::Variable(variable) => Type::Variable(variable),
        }
    }
}

impl<Outer: TypeRef> Display for Type<Outer> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Type::Integer => write!(f, "Integer"),
            Type::Function { parameter, body } => write!(f, "({parameter} -> {body})"),
            Type::Variable(variable) => write!(f, "{variable}"),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TypeVariable(pub Arc<String>);

impl TypeVariable {
    pub fn new(value: String) -> Self {
        Self(Arc::new(value))
    }
    pub fn new_from_str(value: &str) -> Self {
        Self::new(value.to_owned())
    }
}

impl Borrow<str> for TypeVariable {
    fn borrow(&self) -> &str {
        self.0.as_ref()
    }
}

impl Display for TypeVariable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.0.fmt(f)
    }
}
