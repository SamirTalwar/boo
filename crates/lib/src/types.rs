#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Integer,
    Function {
        parameter: Option<Box<Type>>,
        body: Option<Box<Type>>,
    },
}
