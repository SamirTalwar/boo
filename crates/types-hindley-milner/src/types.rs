use boo_core::types::{Monotype, Polytype, Type, TypeVariable};

use crate::fresh::FreshVariables;
use crate::subst::Subst;

pub trait Types {
    fn free(&self) -> im::HashSet<TypeVariable>;

    fn substitute(&self, substitutions: &Subst, fresh: &mut FreshVariables) -> Self;
}

impl Types for Type<Monotype> {
    fn free(&self) -> im::HashSet<TypeVariable> {
        match self {
            Type::Integer => im::HashSet::new(),
            Type::Function { parameter, body } => parameter.free().union(body.free()),
            Type::Variable(variable) => im::hashset![variable.clone()],
        }
    }

    fn substitute(&self, substitutions: &Subst, fresh: &mut FreshVariables) -> Self {
        match self {
            Type::Integer => Type::Integer,
            Type::Function { parameter, body } => Type::Function {
                parameter: parameter.substitute(substitutions, fresh),
                body: body.substitute(substitutions, fresh),
            },
            Type::Variable(variable) => match substitutions.get(variable) {
                None => Type::Variable(variable.clone()),
                Some(t) => (*t.0).clone(),
            },
        }
    }
}

impl Types for Monotype {
    fn free(&self) -> im::HashSet<TypeVariable> {
        self.0.free()
    }

    fn substitute(&self, substitutions: &Subst, fresh: &mut FreshVariables) -> Self {
        self.0.substitute(substitutions, fresh).into()
    }
}

impl Types for Polytype {
    fn free(&self) -> im::HashSet<TypeVariable> {
        let quantifiers = self.quantifiers.iter().cloned().collect();
        self.mono.free().relative_complement(quantifiers)
    }

    fn substitute(&self, substitutions: &Subst, fresh: &mut FreshVariables) -> Self {
        let replacements = self
            .quantifiers
            .iter()
            .map(|q| (q.clone(), fresh.next()))
            .collect::<Vec<_>>();
        let new_quantifiers = replacements
            .iter()
            .map(|(_, after)| after.clone())
            .collect();
        let replacements_subst = replacements
            .into_iter()
            .map(|(before, after)| (before, Type::Variable(after).into()))
            .collect();
        Self {
            quantifiers: new_quantifiers,
            mono: self
                .mono
                .substitute(&replacements_subst, fresh)
                .substitute(substitutions, fresh),
        }
    }
}
