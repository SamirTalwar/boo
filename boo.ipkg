package boo

sourcedir = "proofs"

modules =
    Builtins
  , EvaluationReduction
  , Expression
  , Identifier
  , Native
  , Primitive
  , Test
