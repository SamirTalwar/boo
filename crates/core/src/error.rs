use crate::span::Span;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(Debug, Clone, PartialEq, thiserror::Error, miette::Diagnostic)]
pub enum Error {
    #[error("Unexpected token: {token}")]
    #[diagnostic(code(boo::lex::unexpected_token))]
    UnexpectedToken {
        #[label("unexpected token")]
        span: Span,
        token: String,
    },

    #[error("Parse error: expected one of {expected_tokens:?}")]
    #[diagnostic(code(boo::parse::error))]
    ParseError {
        #[label("{}", expected_one_of(expected_tokens))]
        span: Span,
        expected_tokens: Vec<&'static str>,
    },

    #[error("Unknown variable: {name:?}")]
    #[diagnostic(code(boo::interpret::unknown_variable))]
    UnknownVariable {
        #[label("unknown variable")]
        span: Span,
        name: String,
    },

    #[error("Could not apply the function")]
    #[diagnostic(code(boo::interpret::unknown_variable))]
    InvalidFunctionApplication {
        #[label("invalid function")]
        span: Span,
    },
}

fn expected_one_of(strings: &[&str]) -> String {
    match strings {
        [] => "<nothing>".to_string(),
        [a] => format!("expected {}", a),
        [a, b] => format!("expected {} or {}", a, b),
        _ => format!(
            "expected one of {}, or {}",
            strings[0..strings.len() - 1].join(", "),
            strings[strings.len() - 1]
        ),
    }
}
