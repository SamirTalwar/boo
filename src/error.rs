#[derive(Debug, PartialEq)]
pub enum BooError {
    ParseError(peg::error::ParseError<peg::str::LineCol>),
}

impl std::fmt::Display for BooError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ParseError(error) => write!(f, "Parse error: {}", error),
        }
    }
}
