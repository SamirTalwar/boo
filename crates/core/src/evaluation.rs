use crate::ast::Function;
use crate::error::Result;
use crate::expr::Expr;
use crate::primitive::Primitive;

pub trait Evaluator {
    fn evaluate(&self, expr: Expr) -> Result<Evaluated>;
}

/// An evaluation result. This can be either a primitive value or a closure.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Evaluated {
    Primitive(Primitive),
    Function(Function<Expr>),
}

impl std::fmt::Display for Evaluated {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Evaluated::Primitive(x) => x.fmt(f),
            Evaluated::Function(x) => x.fmt(f),
        }
    }
}
