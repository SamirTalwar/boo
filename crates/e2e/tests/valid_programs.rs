use boo::error::Result;
use boo::evaluation::Evaluator;
use boo::types::{Monotype, Type};
use boo::*;
use boo_naive_evaluator::NaiveEvaluator;
use boo_optimized_evaluator::OptimizedEvaluator;

#[test]
fn test_integer() -> Result<()> {
    check_program("integer", "123", Type::Integer.into(), "123")
}

#[test]
fn test_mathematical_operators() -> Result<()> {
    check_program(
        "mathematical_operators",
        "7 + 3 * 5 - 2",
        Type::Integer.into(),
        "20",
    )
}

#[test]
fn test_overriding_precedence() -> Result<()> {
    check_program(
        "overriding_precedence",
        "2 * (3 + 4)",
        Type::Integer.into(),
        "14",
    )
}

#[test]
fn test_function_application() -> Result<()> {
    check_program(
        "function_application",
        "(fn x -> x + x) 9",
        Type::Integer.into(),
        "18",
    )
}

#[test]
fn test_function_application_with_computation() -> Result<()> {
    check_program(
        "function_application_with_computation",
        "(fn x y -> x + y) (8 * 2) (3 * 4)",
        Type::Integer.into(),
        "28",
    )
}

#[test]
fn test_assignment() -> Result<()> {
    check_program(
        "assignment",
        "let seven = 7 in seven",
        Type::Integer.into(),
        "7",
    )
}

#[test]
fn test_assignment_and_use() -> Result<()> {
    check_program(
        "assignment_and_use",
        "let eight = 8 in eight * 3",
        Type::Integer.into(),
        "24",
    )
}

#[test]
fn test_named_function_application() -> Result<()> {
    check_program(
        "named_function_application",
        "let double = fn input -> input + input in double 6",
        Type::Integer.into(),
        "12",
    )
}

#[test]
fn test_named_function_application_nested() -> Result<()> {
    check_program(
        "named_function_application_nested",
        "let double = fn input -> input + input in double (double 4)",
        Type::Integer.into(),
        "16",
    )
}

#[test]
fn test_function_application_with_named_argument() -> Result<()> {
    check_program(
        "function_application_with_named_argument",
        "let value = 99 in (fn wibble -> wibble - 1) value",
        Type::Integer.into(),
        "98",
    )
}

#[test]
fn test_named_function_application_with_named_argument() -> Result<()> {
    check_program(
        "named_function_application_with_named_argument",
        "let negate = fn thing -> 0 - thing in let life = 42 in negate life",
        Type::Integer.into(),
        "-42",
    )
}

#[test]
fn test_closing_over_a_variable() -> Result<()> {
    check_program(
        "closing_over_a_variable",
        "let something = 12 in let add_something = fn target -> target + something in add_something 9",
        Type::Integer.into(),
        "21",
    )
}

#[test]
fn test_polymorphic_let() -> Result<()> {
    check_program(
        "polymorphic_let",
        "let id = fn x -> x in id id id (id 7)",
        Type::Integer.into(),
        "7",
    )
}

#[test]
fn test_pattern_matching_on_integers() -> Result<()> {
    check_program(
        "pattern_matching_on_integers",
        "match 2 { 1 -> 2; 2 -> 3; 3 -> 4; _ -> 0 }",
        Type::Integer.into(),
        "3",
    )
}

#[test]
fn test_pattern_matching_on_functions() -> Result<()> {
    check_program(
        "pattern_matching_on_functions",
        "(match 1 { 1 -> fn x -> 2; _ -> fn x -> x }) 3",
        Type::Integer.into(),
        "2",
    )
}

#[test]
fn test_expression_type_annotations() -> Result<()> {
    check_program(
        "expression_type_annotations",
        "let id_int = fn x -> (x: Integer) in id_int (1 + (2: Integer))",
        Type::Integer.into(),
        "3",
    )
}

fn check_program(
    name: &str,
    program: &str,
    expected_type: Monotype,
    expected_result_str: &str,
) -> Result<()> {
    let ast = parse(program)?.to_core()?;
    insta::with_settings!({ description => program }, {
        insta::assert_debug_snapshot!(name.to_string() + "__parse", ast);
    });

    let expected_result = match parse(expected_result_str)?.to_core()?.take() {
        ast::Expression::Primitive(primitive) => evaluation::Evaluated::Primitive(primitive),
        expression => panic!("Expected result that is not a primitive: {:?}", expression),
    };

    let actual_type = boo_types_hindley_milner::type_of(&ast)?;
    assert_eq!(actual_type, expected_type);

    let mut optimized_evaluator = OptimizedEvaluator::new();
    builtins::prepare(&mut optimized_evaluator)?;
    let mut naive_evaluator = NaiveEvaluator::new();
    builtins::prepare(&mut naive_evaluator)?;

    let efficient_result = optimized_evaluator.evaluate(ast.clone())?;
    assert_eq!(efficient_result, expected_result);

    let naive_result = naive_evaluator.evaluate(ast)?;
    assert_eq!(naive_result, expected_result);

    Ok(())
}
