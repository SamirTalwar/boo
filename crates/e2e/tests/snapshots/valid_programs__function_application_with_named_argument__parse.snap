---
source: crates/e2e/tests/valid_programs.rs
description: let value = 99 in (fn wibble -> wibble - 1) value
expression: ast
---
Expr {
    span: Some(
        Span {
            start: 0,
            end: 49,
        },
    ),
    expression: Assign(
        Assign {
            name: Name(
                "value",
            ),
            value: Expr {
                span: Some(
                    Span {
                        start: 12,
                        end: 14,
                    },
                ),
                expression: Primitive(
                    Integer(
                        Small(
                            99,
                        ),
                    ),
                ),
            },
            inner: Expr {
                span: Some(
                    Span {
                        start: 19,
                        end: 49,
                    },
                ),
                expression: Apply(
                    Apply {
                        function: Expr {
                            span: Some(
                                Span {
                                    start: 19,
                                    end: 42,
                                },
                            ),
                            expression: Function(
                                Function {
                                    parameter: Name(
                                        "wibble",
                                    ),
                                    body: Expr {
                                        span: Some(
                                            Span {
                                                start: 32,
                                                end: 42,
                                            },
                                        ),
                                        expression: Apply(
                                            Apply {
                                                function: Expr {
                                                    span: Some(
                                                        Span {
                                                            start: 32,
                                                            end: 42,
                                                        },
                                                    ),
                                                    expression: Apply(
                                                        Apply {
                                                            function: Expr {
                                                                span: Some(
                                                                    Span {
                                                                        start: 32,
                                                                        end: 42,
                                                                    },
                                                                ),
                                                                expression: Identifier(
                                                                    Operator(
                                                                        "-",
                                                                    ),
                                                                ),
                                                            },
                                                            argument: Expr {
                                                                span: Some(
                                                                    Span {
                                                                        start: 32,
                                                                        end: 38,
                                                                    },
                                                                ),
                                                                expression: Identifier(
                                                                    Name(
                                                                        "wibble",
                                                                    ),
                                                                ),
                                                            },
                                                        },
                                                    ),
                                                },
                                                argument: Expr {
                                                    span: Some(
                                                        Span {
                                                            start: 41,
                                                            end: 42,
                                                        },
                                                    ),
                                                    expression: Primitive(
                                                        Integer(
                                                            Small(
                                                                1,
                                                            ),
                                                        ),
                                                    ),
                                                },
                                            },
                                        ),
                                    },
                                },
                            ),
                        },
                        argument: Expr {
                            span: Some(
                                Span {
                                    start: 44,
                                    end: 49,
                                },
                            ),
                            expression: Identifier(
                                Name(
                                    "value",
                                ),
                            ),
                        },
                    },
                ),
            },
        },
    ),
}
